### PR TITLE
feat(reconciliation): define recurring expenses

### LINKED ISSUE #105: Define recurring expenses
Allow users to define their recurring monthly expenses, which are used in reconciliation target calculations.

**Expense types**
- **Statement**: a bill paid from a fixed statement balance each month (e.g. credit card)
- **Running balance**: an ongoing balance that persists and accumulates month-to-month (e.g. loan payment)

**Acceptance criteria**
- A "New expense" action opens a creation form
- Fields: name (required), type (statement or running balance, required), typical monthly amount (required, positive)
- Submitting persists the expense via the service layer
- All user-facing strings in a co-located copy file
- Storybook story covers the creation form for each expense type

### PR DESCRIPTION
**Pull request by rmartz**
Adds the ability to define recurring monthly expenses (statement and running balance types) used in reconciliation target calculations. A "New expense" button in the reconcile page header opens a creation dialog with name, type, and typical monthly amount fields; on submit the expense is persisted via `createReconciliationExpense`.

`CreateExpenseDialog` follows the presentational split pattern (`CreateExpenseDialogView` + stateful container) for testability. `ReconcileView` receives an optional `onNewExpense` callback prop and conditionally renders the button; the reconcile page is now a client component that manages dialog open state and wires the submit handler to the service layer.

## Tests
17 tests added, all passing.

Closes #105

---
*Created by Claude Sonnet 4.5*

### CODE REVIEW

**Review by rmartz — COMMENTED** [review]
## Requested changes

- **`ReconcileView.stories.tsx` is not updated for the new prop.** The PR adds an optional `onNewExpense` prop to `ReconcileView` that controls whether the "New expense" button is rendered, but the existing stories (`Desktop`, `Mobile`) are both rendered without any args and will never show the button. The CLAUDE.md requires updating stories to cover key visual states when modifying a UI component. Add at least one story that passes `onNewExpense` so the button-visible state is represented in Storybook alongside the existing no-button state.

- **PR description contains an acceptance-criteria checklist.** The CLAUDE.md specifies that PR descriptions must be descriptive prose — not a task checklist. The prose section at the top is good and should be kept. The `## Acceptance Criteria` checkbox block duplicates information already tracked in the linked issue (#105) and should be removed from the PR body. Issue acceptance criteria are tracked on the issue; the PR body should explain what the PR does, the technical approach, and any non-obvious decisions.

## Verdict

Two items to address before merge: a missing Storybook story state and a PR description cleanup.

---
*Claude Sonnet 4.6*

**Review by rmartz — COMMENTED** [review]
## Prior items

- ✅ **Missing Storybook story for button-visible state** — `WithNewExpenseButton` story added to `ReconcileView.stories.tsx` in the fix-review commit.
- ✅ **PR description contains acceptance-criteria checklist** — checklist removed; description is now descriptive prose.

## Requested changes

- **Empty-uid mutation risk in `page.tsx`** ([page.tsx line 23](https://github.com/rmartz/personal-budget/pull/177#discussion_r3253034830)): `uid` is derived as `user?.uid ?? ""` at render time, before auth may have finished loading. If the user opens and submits the dialog during that window, `createReconciliationExpense` is called with `""` and writes to an invalid shared RTDB path. Guard the page on `authLoading || !user` before exposing the dialog open action, consistent with other authenticated pages in the codebase.

- **Type-toggle buttons lack accessibility semantics** ([CreateExpenseDialog.tsx line 122](https://github.com/rmartz/personal-budget/pull/177#discussion_r3253034840)): The two expense-type buttons communicate selection only through visual styling, with no `aria-pressed` attribute or radio group semantics. Screen reader users will hear two ordinary buttons with no indication of which type is currently selected. Add `aria-pressed={type === ...}` to each button, or restructure the control as a `role="radiogroup"` with `role="radio"` children.

- **`parseFloat` accepts `Infinity` as a valid amount** ([CreateExpenseDialog.tsx line 261](https://github.com/rmartz/personal-budget/pull/177#discussion_r3253034844)): The validation at line 261 checks `isNaN(parsed) || parsed <= 0`, but `parseFloat("1e309")` returns `Infinity`, which is neither `NaN` nor `<= 0`. This allows a non-finite value to pass validation and be submitted to the persistence layer. Add `!isFinite(parsed)` to the condition to reject these values.

- **Container validation logic has no test coverage** ([CreateExpenseDialog.tsx line 253](https://github.com/rmartz/personal-budget/pull/177#discussion_r3253034856)): The `CreateExpenseDialog` container's `handleSubmit` validates name and amount before calling `onSubmit`, but there are no tests exercising those paths — empty name, invalid/zero amount, or `onSubmit` not called when validation fails. The existing integration test only covers the happy path. Add tests for the validation-failure cases in the container so a regression in validation logic would be caught.

- **New conditional button has no test coverage in `ReconcileView.spec.tsx`** ([ReconcileView.tsx line 67](https://github.com/rmartz/personal-budget/pull/177#discussion_r3253034850)): `ReconcileView` now conditionally renders the "New expense" button based on `onNewExpense`, but `ReconcileView.spec.tsx` has no test for this behavior. Add at least two tests: one verifying the button is absent when `onNewExpense` is not provided, and one verifying the button is present and invokes the callback on click.

- **Test name misleads on what is asserted** ([CreateExpenseDialog.spec.tsx line 188](https://github.com/rmartz/personal-budget/pull/177#discussion_r3253034859)): The test at line 183 is named "disables submit and cancel while isSubmitting" but only asserts the submit/creating button is disabled. The cancel button is not asserted, so it could become enabled during submission without the test failing. Either add an assertion for the cancel button's `disabled` state, or narrow the test name to accurately reflect what is checked.

- **Integration test name overstates service-layer wiring** ([CreateExpenseDialog.spec.tsx line 193](https://github.com/rmartz/personal-budget/pull/177#discussion_r3253034867)): The describe block says "submitting persists via service layer" but the test only verifies that the dialog calls its `onSubmit` prop with the correct arguments — `createReconciliationExpense` is not mocked or called. Rename the describe block to accurately describe the scope (e.g. "calls onSubmit prop with form values"), or add a separate test that mocks and verifies the service call.

- **Story args hard-code copy strings that already exist in the copy file** ([CreateExpenseDialog.stories.tsx line 51](https://github.com/rmartz/personal-budget/pull/177#discussion_r3253034863)): The `WithValidationErrors` story sets `nameError` and `typicalAmountError` to literal strings that duplicate the constants in `CreateExpenseDialog.copy.ts`. Per CLAUDE.md, user-facing strings must be sourced from the copy file; reference `CREATE_EXPENSE_DIALOG_COPY.nameRequiredError` and `CREATE_EXPENSE_DIALOG_COPY.amountInvalidError` instead.

## Verdict

Eight Copilot-flagged issues all remain open — none were addressed by the fix-review commit. The most critical is the auth-loading race condition that can write to an invalid Firebase path; the others are accessibility, validation correctness, test coverage, and copy sourcing gaps. All must be resolved before merge.

---
*Claude Sonnet 4.6*

**Review by rmartz — COMMENTED** [review]
## Prior items

- ✅ **Empty-uid mutation risk in `page.tsx`** — `authLoading` is now extracted, `handleCreateExpense` guards with `if (authLoading || !user) return`, and `onNewExpense` is set to `undefined` while auth is loading.
- ✅ **Type-toggle buttons lack accessibility semantics** — `aria-pressed` added to both expense-type buttons in `CreateExpenseDialog.tsx`.
- ✅ **`parseFloat` accepts `Infinity` as a valid amount** — `!isFinite(parsed)` added to the validation condition.
- ✅ **Container validation logic has no test coverage** — Two new tests added for empty name and invalid amount cases in the container integration describe block.
- ✅ **New conditional button has no test coverage in `ReconcileView.spec.tsx`** — Three new tests cover the button-present, button-absent, and click-handler cases.
- ✅ **Test name misleads on what is asserted** — Renamed to "disables the submit button while isSubmitting".
- ✅ **Integration test name overstates service-layer wiring** — Describe block renamed to "CreateExpenseDialog — container integration" with an accurate test name.
- ✅ **Story args hard-code copy strings** — `CREATE_EXPENSE_DIALOG_COPY.nameRequiredError` and `.amountInvalidError` now used in `WithValidationErrors` story args.

## Requested changes

- **Prettier formatting failure in `ReconcileView.spec.tsx`**: The Format CI check is failing on the current HEAD ([run logs](https://github.com/rmartz/personal-budget/actions/runs/25967206652/job/76332695465)) because `src/components/reconcile/ReconcileView.spec.tsx` does not pass `prettier --check`. The file was modified by the fix-review commit (`abdb125`) — the new test block added at the end of the file has formatting that Prettier rejects. Run `pnpm format` (or `pnpm exec prettier --write src/components/reconcile/ReconcileView.spec.tsx`) locally, verify the change, and commit before pushing again.

## Verdict

All eight prior items are resolved. One new issue introduced by the fix-review commit: a Prettier formatting violation in `ReconcileView.spec.tsx` that is failing CI. Fix the formatting and push; the PR should be clean after that.

---
*Claude Sonnet 4.6*

**Review by rmartz — COMMENTED** [review]
## Prior items

- ✅ **Prettier formatting failure in `ReconcileView.spec.tsx`** — format-fix commit `399cc71` collapses the multi-line `expect()` call to a single line; Format CI passes on the current HEAD.

## Verdict

All prior items resolved and no new issues — ready to merge.

---
*Claude Sonnet 4.6*

**Review by rmartz — COMMENTED** [review]
## Requested changes

- **Auth-to-dialog race condition: silent no-op on submit** ([page.tsx line 16](https://github.com/rmartz/personal-budget/pull/177#discussion_r3253451680)): `handleCreateExpense` guards against unauthenticated state with an early `return`, but the returned `Promise<void>` resolves successfully — the dialog's container interprets the resolved promise as a success and closes the dialog as if the expense was persisted. If the user opens the dialog while authenticated and auth state lapses before submit, their input is silently discarded with no error shown. Either throw (or return a rejected promise) from `handleCreateExpense` when the auth guard fires, or close the dialog reactively when `authLoading || !user` becomes true while it is already open.

- **Reconcile page renders without an auth guard** ([page.tsx line 34](https://github.com/rmartz/personal-budget/pull/177#discussion_r3253451681)): The page renders the full `ReconcileView` while `authLoading` is true or `user` is null, whereas other authenticated app pages (e.g. `accounts/page.tsx`, `annuities/page.tsx`) return `null` before rendering protected content. While the current `ReconcileView` only shows placeholder data, the inconsistency creates a pattern mismatch and a footgun for future code that may display real user data. Add an `if (authLoading || !user) return null;` guard before the `return (` statement, consistent with the rest of the app shell.

- **"Type" toggle group label is not programmatically associated with its buttons** ([CreateExpenseDialog.tsx line 117](https://github.com/rmartz/personal-budget/pull/177#discussion_r3253451684)): The visible "Type" label is rendered as a `<p>` element with no `id`, and the two toggle buttons (which correctly have `aria-pressed`) are not wrapped in a labelled group. Screen readers will announce each button's pressed state but will not associate them with the "Type" field label. Wrap the label and buttons in a `<div role="group" aria-labelledby={typeGroupLabelId}>` and assign that id to the `<p>`, or use a `<fieldset>`/`<legend>` pattern.

## Verdict

Three Copilot-flagged issues remain open from the Copilot review posted at 2026-05-16T16:45:36Z (after the last human review). The most critical is the auth-to-dialog race condition that silently discards submitted data; the others are a rendering guard consistency issue and an accessibility gap in the type toggle. All three must be resolved before merge.

---
*Claude Sonnet 4.6*

File: src/components/reconcile/CreateExpenseDialog.tsx [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+
+            {/* Type toggle */}
+            <div className="flex flex-col gap-1.5">
+              <p
+                id={typeLabelId}
+                className="text-sm font-medium leading-none"
+              >
+                {CREATE_EXPENSE_DIALOG_COPY.typeLabel}
+              </p>
+              <div role="group" aria-labelledby={typeLabelId} className="flex gap-2">
```
[the remark this answers is not in the document]
Comment (rmartz) [reply src/components/reconcile/CreateExpenseDialog.tsx]: Fixed — a subsequent commit (`4224317`) ran `pnpm format`; the JSX is now correctly formatted.

**Review by rmartz — COMMENTED** [review]
## Prior items

- ✅ **Auth-to-dialog race/silent no-op** — fixed in `77f368e` (throws when auth guard fires) and confirmed passing CI.
- ✅ **Reconcile page auth guard** — `if (authLoading || !user) return null` added in `77f368e`.
- ✅ **Type toggle group `aria-labelledby` association** — `role="group"` + `aria-labelledby` wired in `77f368e`.
- ✅ **Async submit success/rejection paths not covered** ([CreateExpenseDialog.tsx line 286](https://github.com/rmartz/personal-budget/pull/177#discussion_r3255021363)) — fix-review commit `189c72b` adds exactly the two tests Copilot requested: "calls onOpenChange with false and resets fields after a successful submit" (asserts `onOpenChange(false)`) and "shows submit error and does not call onOpenChange when onSubmit rejects" (asserts `submitError` text and confirms `onOpenChange` not called).
- ❌ **Positive-amount validation branch not directly exercised** ([CreateExpenseDialog.tsx line 273](https://github.com/rmartz/personal-budget/pull/177#discussion_r3255021377)) — the fix-review commit `189c72b` did not add a test for `0` or negative amounts. The existing invalid-amount test (line 251) only submits with an empty amount field, exercising the `isNaN` branch but not the `parsed <= 0` branch. A test submitting `"0"` or `"-5"` is still missing.

## Requested changes

- **Positive-amount validation branch (0 and negative) has no test coverage**: The container test for invalid amount only fires with an empty string, which hits the `isNaN(parsed)` path. The `parsed <= 0` path — covering submitted values of `"0"` or `"-5"` — is not exercised, so a regression removing that check would pass the test suite. Add a test that sets the amount field to `"0"` and confirms `amountInvalidError` is shown and `onSubmit` is not called. A separate test for a negative value (e.g. `"-5"`) is optional but would more directly pin the `<= 0` boundary.

## Verdict

One prior item still open: the zero/negative amount test gap. All other items — including the async submit outcome tests — are resolved.

---
*Claude Sonnet 4.6*

**Review by rmartz — COMMENTED** [review]
## Prior items

- ✅ **Positive-amount validation branch (0 and negative) has no test coverage** — commit `7d658af` adds two new tests: zero amount ("0") and negative amount ("-5"), directly exercising the `parsed <= 0` path.

## Requested changes

- **[`parseFloat` accepts partial numeric prefixes](https://github.com/rmartz/personal-budget/pull/177#discussion_r3255176446)** ([CreateExpenseDialog.tsx line 272](https://github.com/rmartz/personal-budget/pull/177#discussion_r3255176446)): `parseFloat("120abc")` returns `120`, which is not `NaN`, is positive, and is finite — so it passes all three checks and gets submitted to the persistence layer as a valid amount. An input like `"3e"` similarly parses to `3`. Replace `parseFloat` with `Number()` (which returns `NaN` for any non-purely-numeric string) or read `valueAsNumber` from the input element, so malformed strings are rejected at the validation step rather than silently coerced.

- **[Test name overstates what is asserted](https://github.com/rmartz/personal-budget/pull/177#discussion_r3255176455)** ([CreateExpenseDialog.spec.tsx line 331](https://github.com/rmartz/personal-budget/pull/177#discussion_r3255176455)): The test at line 331 is named "calls onOpenChange with false and resets fields after a successful submit", but the body only waits for `onOpenChange` to have been called with `false` — it never asserts that the name, amount, or type fields return to their default values. Either add assertions that confirm the fields are reset (e.g., re-open the dialog and verify the inputs are empty/default), or narrow the test name to "calls onOpenChange with false after a successful submit" so it accurately describes what is checked.

## Verdict

One prior item resolved. Two Copilot-flagged issues remain open: a correctness bug in amount validation that allows partial-prefix numeric strings through, and a misleading test name. Both must be addressed before merge.

---
*Claude Sonnet 4.6*

**Review by rmartz — COMMENTED** [review]
## Prior items

- ✅ **`parseFloat` accepts partial numeric prefixes** — replaced with `Number()` in `CreateExpenseDialog.tsx` line 271; `"120abc"` and `"3e"` now correctly return `NaN` and fail validation.
- ✅ **Test name overstates what is asserted** — renamed to `"calls onOpenChange with false after a successful submit"`, accurately reflecting that only the `onOpenChange(false)` call is asserted.

## Verdict

All prior items resolved and no new issues — ready to merge.

---
*Claude Sonnet 4.6*

**Review by rmartz — COMMENTED** [review]
Requesting Copilot re-review of the fix commit.

**Review by rmartz — COMMENTED** [review]
## Requested changes

- **Alphabetical ordering violation in `ReconcileView.copy.ts`**: The `RECONCILE_VIEW_COPY` `as const` object does not maintain alphabetical key order. `actionNewExpense` is inserted between `actionApply` and `actionConfirmed`, but `C < N` alphabetically — `actionConfirmed` should precede `actionNewExpense`. Per CLAUDE.md, `as const` object keys must be kept in alphabetical order by convention. Move `actionNewExpense` to appear after `actionConfirmed`.

## Verdict

One ordering violation caught by the full-diff pass — not flagged in prior incremental reviews because each pass only showed the insertion context, not the alphabetical relationship between surrounding keys. Fix the key order and this PR is ready to merge.

---
*Posted by Claude Sonnet 4.5*

**Review by rmartz — COMMENTED** [review]
## Prior items

- ✅ **Alphabetical ordering violation in `ReconcileView.copy.ts`** — commit `dfcda46` swaps `actionNewExpense` to appear after `actionConfirmed`, restoring correct alphabetical order. Fixed.

## Requested changes

- **`CreateExpenseDialog.spec.tsx` exceeds the test file split threshold** ([CreateExpenseDialog.spec.tsx line 193](https://github.com/rmartz/personal-budget/pull/177#discussion_r3255260797)): The spec is 389 lines, which exceeds the repository's 360-line split threshold. The CLAUDE.md specifies that test files exceeding this should be split into a `{module}-tests/` directory organized by domain. Since the file contains distinct view tests and container integration tests, split them into `CreateExpenseDialog-tests/` with separate files for each concern (e.g. `CreateExpenseDialogView.spec.tsx` and `CreateExpenseDialogContainer.spec.tsx`).

- **`CreateExpenseDialog.tsx` exceeds the source file split threshold** ([CreateExpenseDialog.tsx line 53](https://github.com/rmartz/personal-budget/pull/177#discussion_r3255260805)): The component is 309 lines, exceeding the 240-line split threshold. The file already uses the presentational split pattern (the PR description notes `CreateExpenseDialogView` + stateful container), but both are co-located in the same file. Per CLAUDE.md, a sub-component must be in its own file when it owns its own stories or tests — and `CreateExpenseDialogView` is directly referenced from stories and tests. Extract `CreateExpenseDialogView` into `CreateExpenseDialogView.tsx`, keeping the stateful `CreateExpenseDialog` container in `CreateExpenseDialog.tsx`.

## Verdict

The alphabetical ordering fix is clean. Two file-length violations introduced by this PR remain: the spec file needs splitting and the view/container need to be separated into distinct modules. Both must be resolved before merge.

---
*Claude Sonnet 4.6*

**Review by rmartz — COMMENTED** [review]
## Prior items

- ✅ **`CreateExpenseDialog.tsx` exceeds the source file split threshold** — fix-review commit `da2d725` extracts `CreateExpenseDialogView` into `CreateExpenseDialogView.tsx` (214 lines) and reduces `CreateExpenseDialog.tsx` to 103 lines (container only). The open Copilot inline thread at [CreateExpenseDialog.tsx line 39](https://github.com/rmartz/personal-budget/pull/177#discussion_r3255293847) noted the same concern and has been resolved.
- ✅ **`CreateExpenseDialog.spec.tsx` exceeds the test file split threshold** — original 389-line spec deleted; replaced by `CreateExpenseDialog-tests/CreateExpenseDialog.spec.tsx` (207 lines, container integration tests) and `CreateExpenseDialog-tests/CreateExpenseDialogView.spec.tsx` (190 lines, view tests). Both are well within the 360-line split threshold.

## Verdict

All prior items resolved and no new issues — ready to merge.

---
*Claude Sonnet 4.6*

**Comment by rmartz** [comment]
**Pre-merge check**

- ✅ No WIP or draft
- ✅ PR approved, no changes requested, no open threads
- ✅ UAT complete
- ✅ Branch up to date, no conflicts
- ✅ All CI green

Merging.

---
*Posted by Claude Sonnet 4.5*
